src/utilities/activities.ts

```diff
@@ -163,3 +167,82 @@ export function preprocessActivityDirectiveDB(
   }
   return { ...activityDirectiveDB, start_time_ms };
 }
+
+export function copyActivityDirectivesToClipboard(sourcePlan: Plan, activities: ActivityDirective[]) {
+  const copiedActivityIds = new Set(activities.map(a => a.id));
+  const clippedActivities = activities.map(activity => {
+    const anchorInSelection = activity.anchor_id !== null && copiedActivityIds.has(activity.anchor_id);
+    return {
+      anchor_id: anchorInSelection ? activity.anchor_id : null,
+      anchored_to_start: activity.anchored_to_start,
+      arguments: activity.arguments,
+      id: activity.id,
+      name: activity.name,
+      start_offset: activity.start_offset,
+      start_time_ms: activity.start_time_ms,
+      tags: activity.tags,
+      type: activity.type,
+    };
+  });
+
+  const clipboard = {
+    activities: clippedActivities,
+    sourcePlanId: sourcePlan.id,
+  };
+
+  setClipboardContent(JSON.stringify(clipboard));
+  showSuccessToast(`Copied ${activities.length} Activity Directive${activities.length ? '' : 's'}`);
+}
+
+export function canPasteActivityDirectivesFromClipboard(destinationPlan: Plan | null): boolean {
+  if (destinationPlan === null) {
+    return false;
+  }
+
+  const serializedClipboard = getClipboardContent();
+  if (serializedClipboard === null) {
+    return false;
+  }
+
+  const clipboard = JSON.parse(serializedClipboard);
+  //current scope, allows copy/paste in the same plan
+  if (clipboard.sourcePlanId !== destinationPlan.id) {
+    return false;
+  }
+
+  return true;
+}
+
+export function pasteActivityDirectivesFromClipboard(
+  destinationPlan: Plan,
+  user: User | null,
+  pasteStartingAtTime?: number,
+) {
+  const serializedClipboard = getClipboardContent();
+  if (serializedClipboard !== null) {
+    const clipboard = JSON.parse(serializedClipboard);
+    const activities: ActivityDirective[] = clipboard.activities;
+
+    //transpose in time if we're given a time, otherwise it paste at the current time
+    if (pasteStartingAtTime !== undefined) {
```

Comment:
Might be stronger to check `typeof pasteStartingAtTime === 'number'`?

Reply:
Sounds good!